Desicion Tree!! -360DIGITMG

Decision Tree is a Supervised learning technique that can be used for both classification and Regression problems.
Decision Tree uses a flowchart like a tree structure to show the predictions that result from a series of feature-based splits. It starts with a root node and ends with a decision made by leaves.

Random Forest
A random forest is a machine learning technique that’s used to solve regression and classification problems. It utilizes ensemble learning, which is a technique that combines many classifiers to provide solutions to complex problems.

A random forest algorithm consists of many decision trees. The ‘forest’ generated by the random forest algorithm is trained through bagging or bootstrap aggregating.

Decision Tree problem 2: Divide the data (Diabetes) into training and test datasets and create a Random Forest Model to classify 'Class Variable'.

Business Objective: Predict whether a patient had diabetes or Not

In [1]:
import pandas as pd               # for Data Manipulation
import numpy as np                #for Mathematical calculations
# Importing company dataset using pandas
diabetes = pd.read_csv(r"C:\Users\D\Desktop\New Assignments  Keys\Datasets\Diabetes.csv")

In [2]:
diabetes

Number of times pregnant   Plasma glucose concentration  \
0                            6                            148   
1                            1                             85   
2                            8                            183   
3                            1                             89   
4                            0                            137   
..                         ...                            ...   
763                         10                            101   
764                          2                            122   
765                          5                            121   
766                          1                            126   
767                          1                             93   

      Diastolic blood pressure   Triceps skin fold thickness  \
0                           72                            35   
1                           66                            29   
2                           64                             0   
3                           66                            23   
4                           40                            35   
..                         ...                           ...   
763                         76                            48   
764                         70                            27   
765                         72                            23   
766                         60                             0   
767                         70                            31   

      2-Hour serum insulin   Body mass index   Diabetes pedigree function  \
0                        0              33.6                        0.627   
1                        0              26.6                        0.351   
2                        0              23.3                        0.672   
3                       94              28.1                        0.167   
4                      168              43.1                        2.288   
..                     ...               ...                          ...   
763                    180              32.9                        0.171   
764                      0              36.8                        0.340   
765                    112              26.2                        0.245   
766                      0              30.1                        0.349   
767                      0              30.4                        0.315   

      Age (years)  Class variable  
0              50             YES  
1              31              NO  
2              32             YES  
3              21              NO  
4              33             YES  
..            ...             ...  
763            63              NO  
764            27              NO  
765            30              NO  
766            47             YES  
767            23              NO  

[768 rows x 9 columns]

## EDA, Pre-processing

In [3]:
diabetes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0    Number of times pregnant      768 non-null    int64  
 1    Plasma glucose concentration  768 non-null    int64  
 2    Diastolic blood pressure      768 non-null    int64  
 3    Triceps skin fold thickness   768 non-null    int64  
 4    2-Hour serum insulin          768 non-null    int64  
 5    Body mass index               768 non-null    float64
 6    Diabetes pedigree function    768 non-null    float64
 7    Age (years)                   768 non-null    int64  
 8    Class variable                768 non-null    object 
dtypes: float64(2), int64(6), object(1)
memory usage: 54.1+ KB


In [4]:
diabetes.iloc[:,-1].value_counts(normalize = True)

NO     0.651042
YES    0.348958
Name:  Class variable, dtype: float64

In [5]:
# converting class vaiable to numerical
diabetes.iloc[:,-1] = diabetes.iloc[:,-1].map({'NO':0,'YES':1})

In [6]:
diabetes.describe()

Number of times pregnant   Plasma glucose concentration  \
count                 768.000000                     768.000000   
mean                    3.845052                     120.894531   
std                     3.369578                      31.972618   
min                     0.000000                       0.000000   
25%                     1.000000                      99.000000   
50%                     3.000000                     117.000000   
75%                     6.000000                     140.250000   
max                    17.000000                     199.000000   

        Diastolic blood pressure   Triceps skin fold thickness  \
count                 768.000000                    768.000000   
mean                   69.105469                     20.536458   
std                    19.355807                     15.952218   
min                     0.000000                      0.000000   
25%                    62.000000                      0.000000   
50%                    72.000000                     23.000000   
75%                    80.000000                     32.000000   
max                   122.000000                     99.000000   

        2-Hour serum insulin   Body mass index   Diabetes pedigree function  \
count             768.000000        768.000000                   768.000000   
mean               79.799479         31.992578                     0.471876   
std               115.244002          7.884160                     0.331329   
min                 0.000000          0.000000                     0.078000   
25%                 0.000000         27.300000                     0.243750   
50%                30.500000         32.000000                     0.372500   
75%               127.250000         36.600000                     0.626250   
max               846.000000         67.100000                     2.420000   

        Age (years)   Class variable  
count    768.000000       768.000000  
mean      33.240885         0.348958  
std       11.760232         0.476951  
min       21.000000         0.000000  
25%       24.000000         0.000000  
50%       29.000000         0.000000  
75%       41.000000         1.000000  
max       81.000000         1.000000

In [7]:
# Missing values
diabetes.isnull().sum()

 Number of times pregnant        0
 Plasma glucose concentration    0
 Diastolic blood pressure        0
 Triceps skin fold thickness     0
 2-Hour serum insulin            0
 Body mass index                 0
 Diabetes pedigree function      0
 Age (years)                     0
 Class variable                  0
dtype: int64

In [8]:
# Input and Output Split
predictors = diabetes.iloc[:, :-1]
target = diabetes.iloc[:,-1]

In [9]:
# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(predictors, target, test_size = 0.25, random_state=0)

In [10]:
from sklearn.tree import DecisionTreeClassifier as DT
model = DT(criterion = 'entropy')
model.fit(x_train, y_train)

DecisionTreeClassifier(criterion='entropy')

In [11]:
# Prediction on Test Data
preds = model.predict(x_test)
pd.crosstab(y_test, preds, rownames=['Actual'], colnames=['Predictions'])

Predictions    0   1
Actual              
0            108  22
1             22  40

In [12]:
# Test Data Accuracy 
from sklearn.metrics import accuracy_score, confusion_matrix
accuracy_score(y_test, model.predict(x_test))

0.7708333333333334

In [13]:
# Prediction on Train Data
preds = model.predict(x_train)
pd.crosstab(y_train, preds, rownames = ['Actual'], colnames = ['Predictions'])

Predictions    0    1
Actual               
0            370    0
1              0  206

In [14]:
# Train Data Accuracy 
from sklearn.metrics import accuracy_score, confusion_matrix
accuracy_score(y_train, model.predict(x_train))

1.0

Test data accuracy is 0.78125 and train data accuracy is 1 so the model is overfitting

In [15]:
from sklearn.ensemble import RandomForestClassifier

In [16]:
rf_clf = RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=42)

In [17]:
rf_clf.fit(x_train, y_train)

RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=42)

In [18]:
# Test Data Accuracy 
from sklearn.metrics import accuracy_score, confusion_matrix
confusion_matrix(y_test, rf_clf.predict(x_test))
accuracy_score(y_test, rf_clf.predict(x_test))

0.8020833333333334

Test data accuracy for decision tree was 0.78 and for Random forest is 0.80, accuracy has been increased by 2%.

In [19]:
# Train Data Accuracy
accuracy_score(y_train, rf_clf.predict(x_train))

1.0

Test data accuracy is 0.802 and train accuracy is 1. so, model is overfitting

## Hyperparameter Tuning
As the model is overfitting, we do some hyperparameter tuning to get the desired result

In [20]:
# Creating new model testing with new parameters
forest_new = RandomForestClassifier(n_estimators=100,max_depth=10,min_samples_split=20,criterion='gini')  # n_estimators is the number of decision trees
forest_new.fit(x_train, y_train)

RandomForestClassifier(max_depth=10, min_samples_split=20)

In [21]:
print('Train accuracy: {}'.format(forest_new.score(x_train, y_train)))
print('Test accuracy: {}'.format(forest_new.score(x_test, y_test)))

Train accuracy: 0.8784722222222222
Test accuracy: 0.796875
